### PR TITLE
Setup JSON Server

### LINKED ISSUE #34: Project Upload
# User-Story

As a user

I want to be able to upload projects

# Notes, Links, Material

# ToDos

- [x] Implement routes in client
- [x] add upload functionality to form

### PR DESCRIPTION
**Pull request by kschmitz123**
It's now possible to upload projects (full implementation with mongoDB will follow later).
After the user has uploaded a project , he will be forwarded to the project page.

toDo: include search/filter function, to search for categories and tags

close #34

### CODE REVIEW

File: client/db.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    {
+      "data": {
+        "image": "https://res.cloudinary.com/kschmitz/image/upload/v1606483223/upload/iyvfricfqui4dw2knkcv.jpg",
+
+        "projectTitle": "test",
+        "description": "beschreibung",
+        "tags": "tag, tag, ttag",
+        "category": "macrame"
+      },
+      "tags": ["tag", "tag", "ttag"],
```
Comment (phlgr) [inline client/db.json]: Why are these doubled?
Is one for the image and one for the project itself?

File: client/src/components/UploadProject.js [hunk 2]
```diff
@@ -29,11 +34,18 @@ export default function UploadProject() {
...
+      setLoading(true);
+      const project = await postProject({ data, tags });
+      setLoading(false);
+      if (!previewSource) return;
+      uploadImage(previewSource);
+      history.push(`/projects/${project.id}`);
+    } catch (error) {
+      setErrorMessage(error.message);
+    }
   };
```
Comment (phlgr) [inline client/src/components/UploadProject.js:49]: Maybe outsource this into a hook

File: client/src/pages/DetailsPage.js [hunk 3]
```diff
@@ -1,9 +1,12 @@
-import React from "react";
+import React, { useEffect } from "react";
```
Comment (phlgr) [inline client/src/pages/DetailsPage.js]:
```suggestion
import { useEffect } from "react";
```

File: client/src/pages/DetailsPage.js [hunk 4]
```diff
@@ -16,54 +19,37 @@ const Title = styled.h3`
...
+        setLoading(true);
+        const newProject = await getProjectById(projectId);
+        setProject(newProject);
+        setLoading(false);
+      } catch (error) {
+        setErrorMessage(error.message);
+      }
+    }
+    fetchData();
+  }, [projectId]);
```
Comment (phlgr) [inline client/src/pages/DetailsPage.js:39]: Same as above, maybe a custom Hook would be nice

File: client/src/pages/DetailsPage.js [hunk 5]
```diff
@@ -16,54 +19,37 @@ const Title = styled.h3`
...
-          dolor sit amet.
-        </div>
+        {loading && <div>Loading...</div>}
+        {errorMessage && <p>{errorMessage}</p>}
+        <ImagePreview
+          src={project.data?.image}
+          alt={project.data?.projectTitle}
+        />
+        <Title>{project.data?.projectTitle}</Title>
+        <div>{project.data?.description}</div>
```
Comment (phlgr) [inline client/src/pages/DetailsPage.js]:
```suggestion
{project.data &&
        <ImagePreview
          src={project.data.image}
          alt={project.data.projectTitle}
        />
        <Title>{project.data.projectTitle}</Title>
        <div>{project.data.description}</div>
}
```

File: client/src/utils/api.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
+export async function postProject(project) {
+  const response = await fetch(`http://localhost:5000/projects/`, {
```
Comment (phlgr) [inline client/src/utils/api.js]:
```suggestion
  const response = await fetch(`/projects/`, {
```

File: client/src/utils/api.js [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    method: "POST",
+    headers: { "Content-Type": "application/json" },
+    body: JSON.stringify(project),
+  });
+  const newProject = await response.json();
+  return newProject;
+}
+
+export async function getProjectById(id) {
+  const response = await fetch(`http://localhost:5000/projects/${id}`);
```
Comment (phlgr) [inline client/src/utils/api.js]:
```suggestion
  const response = await fetch(`/projects/${id}`);
```

File: client/src/utils/api.js (hunk 6 above)
Comment (phlgr) [reply client/src/utils/api.js]: Make these paths relative, else it won't work in production

File: client/package.json [hunk 8]
```diff
@@ -15,6 +15,7 @@
     "@testing-library/jest-dom": "^5.11.6",
     "@testing-library/react": "^11.2.1",
     "@testing-library/user-event": "^12.2.2",
+    "json-server": "^0.16.3",
```
Comment (phlgr) [inline client/package.json]: Maybe start the json-server within your server.js as described here: https://www.npmjs.com/package/json-server#module

**Review by phlgr — CHANGES_REQUESTED** [review]

File: client/db.json (hunk 1 above)
Comment (kschmitz123) [reply client/db.json]: i haven't found a way to format the tags with react-hook-form so i added a second, formatted line. But i'm happy about any suggestions on how to solve this :) (didn't work with useSate

**Review by Sina-Sche — COMMENTED** [review]
Awesome! 💯 🚀

File: client/src/components/UploadProject.js [hunk 9]
```diff
@@ -1,4 +1,4 @@
-import React, { useState } from "react";
+import { useState } from "react";
 import { Button } from "./Button";
```
Comment (bwnstck) [inline client/src/components/UploadProject.js]: Components should have a default export:

Look at this: https://github.com/DiscoDevs/DispoDisco/pull/76

**Review by bwnstck — APPROVED** [review]

**Review by SvLowman — APPROVED** [review]
😊

**Pull request merged**